File: web/src/lib/spaces.ts
// Shared spaces and DMs (M6.2, API.md §3 "Shared spaces"): who is in each space, and author
// cards for other accounts' members, which never sync into this account's projection.
import { apiBase } from './sync/device';
import { apiFetch } from './http';
import { sync } from './sync/client';
import type { MemberRow } from './data';

export interface SpaceAccount { id: string; handle: string | null; display_name: string | null; kind: string }
export interface SpaceInfo { id: string; kind: 'internal' | 'shared' | 'dm'; name: string | null; owner_account_id: string; accounts: SpaceAccount[] }
interface Card { id: string; account_id: string; name: string | null; display_name: string | null; pronouns: string | null; color: string | null; sigils: string[]; avatar_blob: string | null }

async function call(method: string, path: string, body?: unknown) {
  const r = await apiFetch(`${apiBase()}${path}`, {
    method,
    headers: { 'content-type': 'application/json', authorization: `Bearer ${sync.device?.session ?? ''}` },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  const text = await r.text();
  const j = text ? JSON.parse(text) : null;
  if (!r.ok) throw new Error(j?.error?.message ?? `HTTP ${r.status}`);
  return j;
}

export async function listSpaces(): Promise<SpaceInfo[]> {
  return (await call('GET', '/spaces')).items;
}

/** Open (or find) the DM with an account; resolves to the space id. */
export async function openDm(accountId: string): Promise<string> {
  return (await call('POST', '/spaces', { kind: 'dm', accounts: [accountId] })).id;
}

export async function createShared(name: string, accounts: string[]): Promise<string> {
  return (await call('POST', '/spaces', { kind: 'shared', name, accounts })).id;
}

export async function addToSpace(spaceId: string, accounts: string[]): Promise<void> {
  await call('POST', `/spaces/${spaceId}/members`, { accounts });
}

export async function leaveSpace(spaceId: string): Promise<void> {
  await call('DELETE', `/spaces/${spaceId}/members/me`);
}

/** Author cards as member rows, ready to merge into a name lookup. */
export async function authorCards(spaceId: string): Promise<{ accounts: SpaceAccount[]; members: MemberRow[] }> {
  const j = await call('GET', `/spaces/${spaceId}/authors`);
  const members = (j.members as Card[]).map(
    (c): MemberRow => ({
      id: c.id,
      name: c.name ?? 'Someone',
      display_name: c.display_name ?? undefined,
      pronouns: c.pronouns ?? undefined,
      color: c.color ?? '#A09184',
      sigils: c.sigils ?? [],
      proxy_tags: [],
      avatar_blob: c.avatar_blob ?? undefined,
      archived: false,
      deleted: false,
    }),
  );
  return { accounts: j.accounts, members };
}

/** What to call a space in the rail: its name, or for a DM the other account. */
export function spaceTitle(s: { id: string; kind: string; name: string }, info: SpaceInfo | undefined, me: string): string {
  if (s.kind === 'internal') return s.name || 'Home';
  if (s.kind === 'dm') {
    const other = info?.accounts.find((a) => a.id !== me);
    return other ? (other.display_name ?? `@${other.handle}`) : 'Direct message';
  }
  return s.name;
}

File: android/app/src/main/kotlin/garden/vayne/chorus/data/Spaces.kt
package garden.vayne.chorus.data

import org.json.JSONArray
import org.json.JSONObject

private fun text(j: JSONObject, key: String): String? =
    if (j.has(key) && !j.isNull(key)) j.optString(key).takeIf { it.isNotBlank() } else null

/** Server directory data that is intentionally absent from this account's sync projection. */
data class SpaceAccount(val id: String, val handle: String?, val displayName: String?) {
    val shownName: String get() = displayName?.takeIf { it.isNotBlank() }
        ?: handle?.takeIf { it.isNotBlank() }?.let { "@$it" } ?: "Someone"
}

data class SpaceInfo(val id: String, val kind: String, val ownerAccountId: String, val accounts: List<SpaceAccount>)
data class ForeignAuthor(val id: String, val accountId: String, val name: String) {
    companion object {
        fun fromJson(j: JSONObject): ForeignAuthor = ForeignAuthor(
            j.getString("id"), j.getString("account_id"),
            text(j, "display_name") ?: text(j, "name") ?: "Someone",
        )
    }
}

object Spaces {
    private fun JSONArray.objects(): List<JSONObject> = (0 until length()).map { getJSONObject(it) }

    fun directory(j: JSONObject): Map<String, SpaceInfo> = j.getJSONArray("items").objects().associate { s ->
        val id = s.getString("id")
        id to SpaceInfo(id, s.getString("kind"), s.getString("owner_account_id"),
            s.getJSONArray("accounts").objects().map(::account))
    }

    fun connected(j: JSONObject): List<SpaceAccount> = sequenceOf("following", "followers")
        .flatMap { direction -> j.getJSONArray(direction).objects().asSequence() }
        .filter { it.optString("status") == "active" }
        .map { account(it.getJSONObject("account")) }
        .distinctBy { it.id }
        .toList()

    fun authors(j: JSONObject): Map<String, ForeignAuthor> = j.getJSONArray("members").objects()
        .map(ForeignAuthor::fromJson).associateBy { it.id }

    private fun account(j: JSONObject) = SpaceAccount(j.getString("id"), text(j, "handle"), text(j, "display_name"))

    suspend fun list(dev: DeviceRecord): Map<String, SpaceInfo> = directory(Api.call("GET", dev.base, "/spaces", null, dev.session))
    suspend fun connections(dev: DeviceRecord): List<SpaceAccount> = connected(Api.call("GET", dev.base, "/follows", null, dev.session))
    suspend fun authorCards(dev: DeviceRecord, spaceId: String): Map<String, ForeignAuthor> =
        authors(Api.call("GET", dev.base, "/spaces/$spaceId/authors", null, dev.session))

    suspend fun createShared(dev: DeviceRecord, name: String, accounts: List<String>): String =
        Api.post(dev.base, "/spaces", JSONObject().put("kind", "shared").put("name", name.trim())
            .put("accounts", JSONArray(accounts)), dev.session).getString("id")

    suspend fun openDm(dev: DeviceRecord, accountId: String): String =
        Api.post(dev.base, "/spaces", JSONObject().put("kind", "dm")
            .put("accounts", JSONArray(listOf(accountId))), dev.session).getString("id")

    suspend fun leave(dev: DeviceRecord, spaceId: String) {
        Api.call("DELETE", dev.base, "/spaces/$spaceId/members/me", null, dev.session)
    }
}
